### PR TITLE
Improvements to Registry Entry Attributes

### PR DESCRIPTION
**Pull request by Leo40Git**
This PR adds the following improvements to the Registry Entry Attributes API:
- Added validation to all values, computed or otherwise
  - (done by trying to encode the value)
- `RegistryEntryAttribute` refactors!
  - Added `getNullable`, for hot code paths
  - Renamed `getValue` to `get`
    - Old name still exists, but is marked for removal
  - Added iteration methods!
    - `keySet` and `tagKeySet`, to get registry entries/tags that have an associated value
    - `entryIterator` and `tagEntryIterator`, to get pairs of registry entries/tags and their associated values
  - Added "value associated with entry" and "value associated with tag" events
    - Only fired for built-in values (`put`) - for data values, use `ResourceLoaderEvents.END_DATA_PACK_RELOAD`
- Merged data and assets holders
  - These were mutually exclusive anyway
- Added flags system to built-in holder
  - Right now, this is only used to exclude computed values (from `DefaultValueProvider`) from the `dump_builtin_attributes` command
- Marked `RegistryExtensions.register(Registry, Identifier, Object, Consumer<BuiltinAttachmentBuilder>)` and `BuiltinAttachmentBuilder` for removal
  - Nobody used these, and they've pretty much been replaced with the direct `RegistryEntryAttachment.put` method

Additionally, this PR also addresses two bugs:
- Fixed tag-based value associations not being cleared on datapack reload
- Fixed tag-based value associations not being cleared on synchronization from server

### CODE REVIEW

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinAttachmentBuilderImpl.java [hunk 1]
```diff
@@ -18,24 +18,20 @@
 
 import org.jetbrains.annotations.ApiStatus;
 
-import net.minecraft.util.registry.Registry;
-
 import org.quiltmc.qsl.registry.attachment.api.RegistryEntryAttachment;
 import org.quiltmc.qsl.registry.attachment.api.RegistryExtensions;
 
 @ApiStatus.Internal
 public final class BuiltinAttachmentBuilderImpl<R> implements RegistryExtensions.BuiltinAttachmentBuilder<R> {
```
Comment (OroArmor) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinAttachmentBuilderImpl.java:26]: Maybe make this class deprecated or remove it? Doesnt do much anymore.

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentHolder.java [hunk 2]
```diff
@@ -134,7 +129,12 @@ public <T> void putValue(RegistryEntryAttachment<R, T> attachment, TagKey<R> tag
 		this.valueTagTable.put(attachment, tag, value);
 	}
 
+	protected <T> void putValueFromTag(RegistryEntryAttachment<R, T> attachment, R entry, T value) {
+		putValue(attachment, entry, value);
+	}
+
 	public void clear() {
 		this.valueTable.clear();
+		this.valueTagTable.clear();
```
Comment (OroArmor) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentHolder.java]: oops

**Review by OroArmor — APPROVED** [review]

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentHolder.java (hunk 2 above)
Comment (Leo40Git) [reply library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentHolder.java]: Hey, at least we noticed it before any end users :P

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 3]
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
...
 	@Override
 	public void put(TagKey<R> entry, V value) {
+		CodecUtils.assertValid(codec, value);
 		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
+		tagValueAddedEvent.invoker().onTagValueAdded(entry, value);
+	}
+
+	@Override
+	public Event<ValueAdded<R, V>> valueAddedEvent() {
+		return valueAddedEvent;
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		return this.valueAddedEvent;
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 4]
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
...
+	}
+
+	@Override
+	public Event<TagValueAdded<R, V>> tagValueAddedEvent() {
+		return tagValueAddedEvent;
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		return this.tagValueAddedEvent;
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 5]
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
...
+		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value,
+				BuiltinRegistryEntryAttachmentHolder.FLAG_NONE);
+		valueAddedEvent.invoker().onValueAdded(entry, value);
 	}
 
 	@Override
 	public void put(TagKey<R> entry, V value) {
+		CodecUtils.assertValid(codec, value);
 		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
+		tagValueAddedEvent.invoker().onTagValueAdded(entry, value);
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		this.tagValueAddedEvent.invoker().onTagValueAdded(entry, value);
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 6]
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
 
 	@Override
 	public void put(R entry, V value) {
-		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
+		CodecUtils.assertValid(codec, value);
+		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value,
+				BuiltinRegistryEntryAttachmentHolder.FLAG_NONE);
+		valueAddedEvent.invoker().onValueAdded(entry, value);
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		this.valueAddedEvent.invoker().onValueAdded(entry, value);
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 7]
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
 
 	@Override
 	public void put(R entry, V value) {
-		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
+		CodecUtils.assertValid(codec, value);
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		CodecUtils.assertValid(this.codec, value);
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 8]
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
...
+		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value,
+				BuiltinRegistryEntryAttachmentHolder.FLAG_NONE);
+		valueAddedEvent.invoker().onValueAdded(entry, value);
 	}
 
 	@Override
 	public void put(TagKey<R> entry, V value) {
+		CodecUtils.assertValid(codec, value);
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		CodecUtils.assertValid(this.codec, value);
```

**Review by LambdAurora — APPROVED** [review]

**Comment by Leo40Git** [comment]
(forgot to add refactor tag, since - yeah, this ~~breaks ABI~~ marks stuff for removal)

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 9]
```diff
@@ -43,6 +46,17 @@ public RegistryEntryAttachmentImpl(Registry<R> registry, Identifier id, Class<V>
 		this.valueClass = valueClass;
 		this.codec = codec;
 		this.side = side;
+
+		valueAddedEvent = Event.create(ValueAdded.class, listeners -> (entry, value) -> {
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		this.valueAddedEvent = Event.create(ValueAdded.class, listeners -> (entry, value) -> {
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java [hunk 10]
```diff
@@ -43,6 +46,17 @@ public RegistryEntryAttachmentImpl(Registry<R> registry, Identifier id, Class<V>
...
+			for (var listener : listeners) {
+				listener.onValueAdded(entry, value);
+			}
+		});
+		tagValueAddedEvent = Event.create(TagValueAdded.class, listeners -> (entry, value) -> {
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java]:
```suggestion
		this.tagValueAddedEvent = Event.create(TagValueAdded.class, listeners -> (entry, value) -> {
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java [hunk 11]
```diff
@@ -199,7 +200,7 @@ static <R> Builder<R, String> stringBuilder(Registry<R> registry, Identifier id)
 	 * @param entry registry entry
 	 * @return attachment value, or empty if no value is assigned
 	 */
-	Optional<V> getValue(R entry);
+	Optional<V> get(R entry);
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java]:
```suggestion
	Optional<V> get(R entry);
	
	/**
	 * Gets the value associated with this attachment for the specified entry.
	 *
	 * @param entry registry entry
	 * @return attachment value, or empty if no value is assigned
	 * @deprecated use {@link #get} instead
	 */
	@Deprecated(forRemoval = true)
	default Optional<V> getValue(R entry) {
		return this.get(entry);
	}
```

**Review by LambdAurora — CHANGES_REQUESTED** [review]

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinRegistryEntryAttachmentHolder.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+	BuiltinRegistryEntryAttachmentHolder() {
+		this.valueFlagTable = Tables.newCustomTable(new Object2ReferenceOpenHashMap<>(), Reference2IntOpenHashMap::new);
+	}
+
+	public int getValueFlags(RegistryEntryAttachment<R, ?> attachment, R entry) {
+		return Objects.requireNonNullElse(this.valueFlagTable.get(attachment, entry), FLAG_NONE);
+	}
+
+	public boolean isValueComputed(RegistryEntryAttachment<R, ?> attachment, R entry) {
+		return (getValueFlags(attachment, entry) & FLAG_COMPUTED) == FLAG_COMPUTED;
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinRegistryEntryAttachmentHolder.java]:
```suggestion
		return (this.getValueFlags(attachment, entry) & FLAG_COMPUTED) == FLAG_COMPUTED;
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/CodecUtils.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.JsonOps;
+
+public final class CodecUtils {
+	private CodecUtils() { }
+
+	public static <T> void assertValid(Codec<T> codec, T value) {
+		var result = codec.encodeStart(JsonOps.INSTANCE, value);
+		if (result.result().isEmpty()) {
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/CodecUtils.java]:
```suggestion

		if (result.result().isEmpty()) {
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ComputedDefaultRegistryEntryAttachmentImpl.java [hunk 14]
```diff
@@ -39,16 +39,29 @@ public ComputedDefaultRegistryEntryAttachmentImpl(Registry<R> registry, Identifi
...
 		if (result.hasFailed()) {
 			COMPUTE_LOGGER.error("Failed to compute value for entry {}: {}", this.registry.getId(entry), result.error());
-			return Optional.empty();
+			return null;
 		} else {
 			var value = result.get();
-			RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
-			return Optional.of(value);
+			var encoded = codec.encodeStart(JsonOps.INSTANCE, value);
+			if (encoded.result().isEmpty()) {
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ComputedDefaultRegistryEntryAttachmentImpl.java:52]:
```suggestion

			if (encoded.result().isEmpty()) {
```

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ConstantDefaultRegistryEntryAttachmentImpl.java [hunk 15]
```diff
@@ -29,11 +29,21 @@
 	public ConstantDefaultRegistryEntryAttachmentImpl(Registry<R> registry, Identifier id, Class<V> valueClass,
 			Codec<V> codec, Side side, V defaultValue) {
 		super(registry, id, valueClass, codec, side);
+
+		var result = this.codec.encodeStart(JsonOps.INSTANCE, defaultValue);
+		if (result.result().isEmpty()) {
```
Comment (LambdAurora) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ConstantDefaultRegistryEntryAttachmentImpl.java]:
```suggestion

		if (result.result().isEmpty()) {
```

**Review by LambdAurora — COMMENTED** [review]
The day I catch everything at once will be a good day.

I don't remember what happened to that ordering stuff we talked earlier though.

**Review by LambdAurora — APPROVED** [review]

**Comment by Leo40Git** [comment]
Moved a critical bugfix to #144.

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java [hunk 16]
```diff
@@ -261,6 +323,102 @@ public boolean shouldLoad(ResourceType source) {
 		}
 	}
 
+	/**
+	 * Specifies a value association entry.
+	 *
+	 * @param <R> type of registry entry
+	 * @param <V> type of value
+	 */
+	@SuppressWarnings("ClassCanBeRecord")
```
Comment (EnnuiL) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java]: I'm curious about why these (and other similar classes) aren't records

File: library/data/registry_entry_attachments/src/testmod/java/org/quiltmc/qsl/registry/attachment/test/DispatchedAttachmentTest.java [hunk 17]
```diff
@@ -91,6 +101,28 @@ public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand han
...
 	@Override
 	public void onInitialize(ModContainer mod) {
-		MODULAR_FUNCTION.put(TagKey.of(Registry.ITEM_KEY, new Identifier("quilt", "modular_tag_1")), new SendMessageFuncValue("Provided via tag"));
+		MODULAR_FUNCTION.put(TagKey.of(Registry.ITEM_KEY, new Identifier("quilt", "modular_tag_1")),
+				new SendMessageFuncValue("Built-in value via tag!"));
+	}
+
+	@Override
+	public void onEndDataPackReload(@Nullable MinecraftServer server, ResourceManager resourceManager, @Nullable Throwable error) {
+		if (error != null) {
```
Comment (EnnuiL) [inline library/data/registry_entry_attachments/src/testmod/java/org/quiltmc/qsl/registry/attachment/test/DispatchedAttachmentTest.java]: Nitpick: perhaps inline the return?

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinRegistryEntryAttachmentHolder.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
Comment (EnnuiL) [inline library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinRegistryEntryAttachmentHolder.java]: Nitpick:
```suggestion
 * Copyright 2022 QuiltMC
```

**Review by Ecorous — APPROVED** [review]

**Review by EnnuiL — COMMENTED** [review]
I just remembered that I've written down a really basic draft of a review and then abandoned it; I don't have any other complaints *and* perhaps this review should be fine

File: library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java (hunk 16 above)
Comment (Leo40Git) [reply library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java]: Honestly I don't remember why I have a vendetta against records either

Comment (Leo40Git) [reply library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java]: nvm, I remembered: you can't document the getters
It's probably fine, though

Comment (EnnuiL) [reply library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java]: That makes kinda sense? Although yeah, I don't really see a point on documenting with the very boilerplate-y "returns the X"; I feel like the comments on the record itself already helps, although you did manage to pull it off anyway, so perhaps this isn't much of an issue anyway

**Pull request merged**